appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java

```diff
@@ -311,54 +319,62 @@ public void refreshTable(boolean needToSort) {
     refreshSortIndicators();
 
     // Refill the table.
-    table.resize(1 + projects.size(), 5);
-    int row = 1;
+    table.resize(2, 5);
+    int row = 0;
     for (Project project : projects) {
-      ProjectWidgets pw = projectWidgets.get(project);
-      if (selectedProjects.contains(project)) {
-        table.getRowFormatter().setStyleName(row, "ode-ProjectRowHighlighted");
-        pw.checkBox.setValue(true);
-      } else {
-        table.getRowFormatter().setStyleName(row, "ode-ProjectRowUnHighlighted");
-        pw.checkBox.setValue(false);
-      }
-      table.setWidget(row, 0, pw.checkBox);
-      table.setWidget(row, 1, pw.nameLabel);
-      table.setWidget(row, 2, pw.dateCreatedLabel);
-      table.setWidget(row, 3, pw.dateModifiedLabel);
-      table.setWidget(row, 4, pw.publishedLabel);
-      if(Ode.getGallerySettings().galleryEnabled()){
-        if (project.isPublished()) {
-          pw.publishedLabel.setText(PUBLISHED);
+      if (project.isInTrash() == isInTrash) {
+        row++;
+        ProjectWidgets pw = projectWidgets.get(project);
+        if (selectedProjects.contains(project)) {
+          table.getRowFormatter().setStyleName(row, "ode-ProjectRowHighlighted");
+          pw.checkBox.setValue(true);
         } else {
-          pw.publishedLabel.setText(NOT_PUBLISHED);
+          table.getRowFormatter().setStyleName(row, "ode-ProjectRowUnHighlighted");
+          pw.checkBox.setValue(false);
+        }
+        pw.checkBox.setName(String.valueOf(row));
+        table.insertRow(row + 1);
+        table.setWidget(row, 0, pw.checkBox);
+        table.setWidget(row, 1, pw.nameLabel);
+        table.setWidget(row, 2, pw.dateCreatedLabel);
+        table.setWidget(row, 3, pw.dateModifiedLabel);
+        table.setWidget(row, 4, pw.publishedLabel);
+        if (Ode.getGallerySettings().galleryEnabled()) {
+          if (project.isPublished()) {
+            pw.publishedLabel.setText(PUBLISHED);
+          } else {
+            pw.publishedLabel.setText(NOT_PUBLISHED);
+          }
         }
       }
-
-      row++;
     }
+    table.resizeRows(row + 1);
```

Comment:
OK, I'm not clear on what's going on under the hood here. 

The total size of the grid was set at line 314 of the original code. Then we loop through each project and populate the rows of the grid. We could do that because active projects and trashed projects were stored in separate but identical structures, so we knew the number of rows just from checking the size.

In the new code, I don't know at the start how large the grid will be because I am referencing the trash status of each project as I loop through. We still loop through the projects and pull the same information, resulting in a grid of the same size.

You're indicating that the setSize reuses existing rows, which wasn't obvious to me but makes sense. I can adjust this code to reuse existing grid rows if performance is a concern. Rather than setting the grid size before the loop, I'll have to check the size of the grid as I loop, add rows if I need them, and truncate anything unused at the end.